File: ios/calorietracker/Insights/RecoveryEngine.swift
import Foundation

// Daily Recovery (0–100): weighted personal-baseline z-scores of last night's signals, a training-load modifier,
// a label and the signals behind it. Port of the "Recovery" section of `scripts/insights_reference.py`.

nonisolated enum RecoveryEngine {
    private static func subscore(_ zDir: Double, _ rc: InsightsConfig.Recovery) -> Double {
        InsightsMath.clamp(rc.subscoreCenter + rc.subscoreSlope * zDir, 0.0, 100.0)
    }

    /// (deviation minutes, sub-score) of last night's midpoint vs the mean midpoint of the prior 14 nights.
    private static func consistency(_ nights: [String: InsightsNight], _ day: String, _ tz: TimeZone,
                                    _ rc: InsightsConfig.Recovery) -> (Double, Double)? {
        var prior: [Double] = []
        for k in stride(from: rc.consistencyWindowDays, to: 0, by: -1) {
            let d = InsightsDay.add(day, -k)
            if let n = nights[d] {
                prior.append(Double(InsightsDay.sleepMidpointMinutes(startMs: n.startMs, endMs: n.endMs, wakeDay: d, tz)))
            }
        }
        guard prior.count >= rc.consistencyMinNights, let today = nights[day] else { return nil }
        let mid = Double(InsightsDay.sleepMidpointMinutes(startMs: today.startMs, endMs: today.endMs, wakeDay: day, tz))
        let dev = abs(mid - InsightsMath.mean(prior))
        return (dev, InsightsMath.clamp(100.0 * (1.0 - dev / rc.consistencyZeroAtMin), 0.0, 100.0))
    }

    /// Morning readiness for `day` (the wake day of last night's sleep). See docs/insights.md.
    static func recovery(_ inputs: InsightsInputs, day: String, config: InsightsConfig) -> RecoveryResult {
        let rc = config.recovery
        let tz = InsightsDay.timeZone(inputs.timeZone)
        let nights = BaselineEngine.validNights(inputs, config: config)
        let sleepSeries = nights.mapValues(\.asleepMin)
        let fallback = Set(inputs.overnightFallback)
        var comps: [RecoveryComponent] = []
        var raws: [String: BaselineEngine.Raw] = [:]
        for c in rc.components {
            guard let m = config.metrics[c.metric] else { continue }
            let s = c.id == "sleep" ? sleepSeries : (inputs.series[c.metric] ?? [:])
            let b = BaselineEngine.baselineRaw(s, day, m)
            var item = RecoveryComponent(
                id: c.id, weight: c.weight, available: false, value: b.recent, baseline: b.mean, delta: b.delta,
                pct: b.pct, z: b.z, subscore: nil, impact: nil, baselineN: b.n, baselineConfidence: b.confidence,
                fallback: fallback.contains(c.metric), consistencyDeviationMin: nil, consistencySubscore: nil)
            if b.isOK, b.recent != nil, let z = b.z {
                var sub: Double
                switch c.mode {
                case "higher": sub = subscore(z, rc)
                case "lower": sub = subscore(-z, rc)
                case "band": sub = subscore(-max(0.0, abs(z) - (c.toleranceZ ?? 0)), rc)
                case "drop_only": sub = subscore(-max(0.0, -z - (c.toleranceZ ?? 0)), rc)
                default:
                    // Sleep: duration vs personal baseline, blended with midpoint consistency when available.
                    sub = subscore(z, rc)
                    if let (dev, cons) = consistency(nights, day, tz, rc) {
                        item.consistencyDeviationMin = dev
                        item.consistencySubscore = cons
                        sub = (c.durationShare ?? 1) * sub + (c.consistencyShare ?? 0) * cons
                    }
                }
                item.available = true
                item.subscore = sub
            }
            comps.append(item)
            raws[c.id] = b
        }
        let sleepB = raws["sleep"]
        let sleepAvailable = comps.first { $0.id == "sleep" }?.available ?? false
        let heart = rc.heartComponents.compactMap { raws[$0] }
        let need = config.metrics["sleep"]?.minPoints ?? 14
        let have = min(sleepB?.n ?? 0, heart.map(\.n).max() ?? 0)
        var out = RecoveryResult(day: day, status: "ok", components: comps, positives: [], negatives: [])
        if !(sleepB?.isOK ?? false) || heart.allSatisfy({ !$0.isOK }) {
            out.status = "collecting"
            out.collecting = InsightsCollecting(have: min(have, need), need: need)
        } else if !sleepAvailable {
            out.status = "no_sleep"
        } else if !comps.contains(where: { rc.heartComponents.contains($0.id) && $0.available }) {
            out.status = "no_heart_data"
        }
        if out.status != "ok" { return rounded(out) }

        var wsum = 0.0
        for i in comps where i.available { wsum += i.weight }
        var score = 0.0
        for idx in comps.indices where comps[idx].available {
            let sub = comps[idx].subscore ?? 0
            score += comps[idx].weight * sub
            comps[idx].impact = (sub - rc.subscoreCenter) * comps[idx].weight / wsum
        }
        score /= wsum
        let loads = TrainingLoadEngine.dailyLoads(inputs.workouts, timeZone: tz, config: config)
        let load = TrainingLoadEngine.loadRaw(loads, InsightsDay.add(day, -1), config: config)
        var mod = 0
        if load.category == "high" {
            let lm = rc.loadModifier
            mod = (load.ratio.map { $0 > lm.veryHighRatio } ?? false) ? lm.veryHigh : lm.high
        }
        let final = min(max(InsightsMath.roundInt(score + Double(mod)), 0), 100)
        let band = rc.bands.first { final >= $0.min } ?? rc.bands[rc.bands.count - 1]
        let avail = comps.filter(\.available)
        let allHigh = avail.allSatisfy { $0.baselineConfidence == "high" }
        let conf = avail.count >= 4 && allHigh ? "high" : (avail.count >= 3 || allHigh ? "medium" : "low")
        var signals: [RecoverySignal] = []
        for i in avail {
            let template = rc.contributors[i.id] ?? i.id
            signals.append(RecoverySignal(id: i.id, impact: i.impact ?? 0,
                                          text: InsightsFormat.fill(template, contributorParams(i))))
        }
        let pos = stableSorted(signals.filter { $0.impact > 0 }) { -$0.impact }
        var neg = stableSorted(signals.filter { $0.impact < 0 }) { $0.impact }
        let loadLabel = config.trainingLoad.labels[load.category] ?? load.category
        if mod < 0 {
            neg.append(RecoverySignal(id: "training_load", impact: Double(mod),
                                      text: InsightsFormat.fill(rc.contributors["training_load"] ?? "", ["category": .text(loadLabel)])))
            neg = stableSorted(neg) { $0.impact }
        }
        out.components = comps
        out.score = final
        out.label = band.id
        out.labelText = band.label
        out.recommendation = band.recommendation
        out.confidence = conf
        out.positives = pos
        out.negatives = neg
        out.load = RecoveryLoad(day: load.day, load: load.load, mean28d: load.mean28d, ratio: load.ratio,
                                category: load.category, label: loadLabel, modifier: mod)
        return rounded(out)
    }

    /// Stable ascending sort by key (ties keep their order).
    static func stableSorted<T>(_ items: [T], by key: (T) -> Double) -> [T] {
        items.enumerated().sorted { a, b in
            let ka = key(a.element), kb = key(b.element)
            return ka != kb ? ka < kb : a.offset < b.offset
        }.map(\.element)
    }

    private static func contributorParams(_ i: RecoveryComponent) -> [String: InsightsParam] {
        let value = i.value ?? 0
        let delta = i.delta ?? 0
        return ["pct": .text(InsightsFormat.signed(i.pct ?? 0.0, 0)), "delta0": .text(InsightsFormat.signed(delta, 0)),
                "delta1": .text(InsightsFormat.signed(delta, 1)),
                "value1": .text(String(format: "%.1f", InsightsMath.roundTo(value, 1))),
                "duration": .text(InsightsFormat.duration(value))]
    }

    private static func rounded(_ result: RecoveryResult) -> RecoveryResult {
        var out = result
        for idx in out.components.indices {
            var i = out.components[idx]
            i.value = InsightsMath.roundTo(i.value, 2)
            i.baseline = InsightsMath.roundTo(i.baseline, 2)
            i.delta = InsightsMath.roundTo(i.delta, 2)
            i.z = InsightsMath.roundTo(i.z, 2)
            i.pct = InsightsMath.roundTo(i.pct, 1)
            i.subscore = InsightsMath.roundTo(i.subscore, 1)
            i.impact = InsightsMath.roundTo(i.impact, 1)
            i.consistencyDeviationMin = InsightsMath.roundTo(i.consistencyDeviationMin, 1)
            i.consistencySubscore = InsightsMath.roundTo(i.consistencySubscore, 1)
            out.components[idx] = i
        }
        out.positives = out.positives.map { var s = $0; s.impact = InsightsMath.roundTo(s.impact, 1); return s }
        out.negatives = out.negatives.map { var s = $0; s.impact = InsightsMath.roundTo(s.impact, 1); return s }
        if var load = out.load {
            load.load = InsightsMath.roundTo(load.load, 1)
            load.mean28d = InsightsMath.roundTo(load.mean28d, 1)
            load.ratio = InsightsMath.roundTo(load.ratio, 2)
            out.load = load
        }
        return out
    }
}

File: android/app/src/main/java/com/ayuvo/health/insights/RecoveryEngine.kt
package com.ayuvo.health.insights

import com.ayuvo.health.insights.InsightsMath.clamp
import com.ayuvo.health.insights.InsightsMath.fill
import com.ayuvo.health.insights.InsightsMath.fmtDuration
import com.ayuvo.health.insights.InsightsMath.fmtSigned
import com.ayuvo.health.insights.InsightsMath.mean
import com.ayuvo.health.insights.InsightsMath.roundInt
import com.ayuvo.health.insights.InsightsMath.roundTo
import com.ayuvo.health.insights.InsightsMath.sleepMidpointMin
import java.time.LocalDate
import java.time.ZoneId
import java.util.Locale
import kotlin.math.abs

/**
 * Daily Recovery, 0–100 (docs/insights.md, "How Recovery is calculated"): each component scores
 * `clamp(50 + 20·z_dir)` against its personal baseline, the available ones are averaged with
 * renormalised weights and a high previous-day training load takes points off. Ported from
 * `recovery` in the reference; [day] is the wake day of last night's sleep.
 */
object RecoveryEngine {

    private class Item(
        val id: String,
        val weight: Double,
        var available: Boolean,
        val value: Double?,
        val baseline: Double?,
        val delta: Double?,
        val pct: Double?,
        val z: Double?,
        var subscore: Double?,
        var impact: Double?,
        val baselineN: Int,
        val baselineConfidence: String,
        val fallback: Boolean,
        var consistencyDeviationMin: Double?,
        var consistencySubscore: Double?
    )

    private fun subscore(zDir: Double, rc: InsightsConfig.Recovery): Double =
        clamp(rc.subscoreCenter + rc.subscoreSlope * zDir, 0.0, 100.0)

    /** (deviation minutes, sub-score) of last night's midpoint vs the mean midpoint of the prior nights. */
    private fun consistency(nights: Map<LocalDate, SleepInput>, day: LocalDate, zone: ZoneId, rc: InsightsConfig.Recovery): Pair<Double, Double>? {
        val prior = ArrayList<Double>()
        for (k in rc.consistencyWindowDays downTo 1) {
            val d = day.minusDays(k.toLong())
            nights[d]?.let { prior += sleepMidpointMin(it, d, zone).toDouble() }
        }
        if (prior.size < rc.consistencyMinNights) return null
        val dev = abs(sleepMidpointMin(nights.getValue(day), day, zone) - mean(prior))
        return dev to clamp(100.0 * (1.0 - dev / rc.consistencyZeroAtMin), 0.0, 100.0)
    }

    fun recovery(inputs: InsightsInputs, day: LocalDate, cfg: InsightsConfig): RecoveryResult {
        val rc = cfg.recovery
        val zone = inputs.timeZone
        val nights = BaselineEngine.validNights(inputs, cfg)
        val sleepSeries = nights.mapValues { it.value.asleepMin!! }
        val items = ArrayList<Item>()
        val raw = HashMap<String, Pair<Item, BaselineEngine.Raw>>()
        for (c in rc.components) {
            val m = cfg.metric(c.metric)
            val s = if (c.id == "sleep") sleepSeries else inputs.series(c.metric)
            val b = BaselineEngine.raw(s, day, m)
            val item = Item(
                id = c.id, weight = c.weight, available = false, value = b.recent, baseline = b.mean, delta = b.delta,
                pct = b.pct, z = b.z, subscore = null, impact = null, baselineN = b.n, baselineConfidence = b.confidence,
                fallback = c.metric in inputs.overnightFallback, consistencyDeviationMin = null, consistencySubscore = null
            )
            if (b.ok && b.recent != null) {
                val z = b.z!!
                val sub = when (c.mode) {
                    "higher" -> subscore(z, rc)
                    "lower" -> subscore(-z, rc)
                    "band" -> subscore(-maxOf(0.0, abs(z) - c.toleranceZ), rc)
                    "drop_only" -> subscore(-maxOf(0.0, -z - c.toleranceZ), rc)
                    else -> {
                        // Sleep: duration vs the personal baseline, blended with midpoint consistency when known.
                        var s0 = subscore(z, rc)
                        consistency(nights, day, zone, rc)?.let { (dev, cons) ->
                            item.consistencyDeviationMin = dev
                            item.consistencySubscore = cons
                            s0 = c.durationShare * s0 + c.consistencyShare * cons
                        }
                        s0
                    }
                }
                item.available = true
                item.subscore = sub
            }
            items += item
            raw[c.id] = item to b
        }
        val (sleepItem, sleepB) = raw.getValue("sleep")
        val heart = rc.heartComponents.map { raw.getValue(it) }
        val need = cfg.metric("sleep").minPoints
        val have = minOf(sleepB.n, heart.maxOf { it.second.n })
        var status = "ok"
        var collecting: Collecting? = null
        if (!sleepB.ok || heart.all { !it.second.ok }) {
            status = "collecting"
            collecting = Collecting(minOf(have, need), need)
        } else if (!sleepItem.available) {
            status = "no_sleep"
        } else if (heart.none { it.first.available }) {
            status = "no_heart_data"
        }
        if (status != "ok") {
            return result(day, status, null, null, collecting, items, emptyList(), emptyList(), null)
        }
        val avail = items.filter { it.available }
        var wsum = 0.0
        for (i in avail) wsum += i.weight
        var score = 0.0
        for (i in avail) {
            score += i.weight * i.subscore!!
            i.impact = (i.subscore!! - rc.subscoreCenter) * i.weight / wsum
        }
        score /= wsum
        val load = TrainingLoad.raw(TrainingLoad.dailyLoads(inputs.workouts, zone, cfg), day.minusDays(1), cfg)
        var mod = 0
        if (load.category == "high") {
            mod = if (load.ratio != null && load.ratio > rc.loadVeryHighRatio) rc.loadVeryHigh else rc.loadHigh
        }
        val final = clamp(roundInt(score + mod).toDouble(), 0.0, 100.0).toInt()
        val band = rc.bands.first { final >= it.min }
        val allHigh = avail.all { it.baselineConfidence == "high" }
        val confidence = if (avail.size >= 4 && allHigh) "high" else if (avail.size >= 3 || allHigh) "medium" else "low"
        val signals = avail.map { RecoverySignal(it.id, it.impact!!, fill(rc.contributors.getValue(it.id), contributorParams(it))) }
        val pos = signals.filter { it.impact > 0 }.sortedBy { -it.impact }
        var neg = signals.filter { it.impact < 0 }.sortedBy { it.impact }
        val loadLabel = cfg.trainingLoad.labels.getValue(load.category)
        if (mod < 0) {
            neg = (neg + RecoverySignal("training_load", mod.toDouble(), fill(rc.contributors.getValue("training_load"), mapOf("category" to loadLabel))))
                .sortedBy { it.impact }
        }
        val recoveryLoad = RecoveryLoad(
            day = load.day, load = roundTo(load.load, 1), mean28d = roundTo(load.mean28d, 1), ratio = roundTo(load.ratio, 2),
            category = load.category, label = loadLabel, modifier = mod
        )
        return result(
            day, "ok", final, band, null, items,
            pos.map { it.copy(impact = roundTo(it.impact, 1)) }, neg.map { it.copy(impact = roundTo(it.impact, 1)) },
            recoveryLoad, confidence
        )
    }

    private fun contributorParams(i: Item): Map<String, Any?> = mapOf(
        "pct" to fmtSigned(i.pct ?: 0.0, 0),
        "delta0" to fmtSigned(i.delta!!, 0),
        "delta1" to fmtSigned(i.delta, 1),
        "value1" to String.format(Locale.ROOT, "%.1f", roundTo(i.value!!, 1)),
        "duration" to fmtDuration(i.value)
    )

    private fun result(
        day: LocalDate,
        status: String,
        score: Int?,
        band: InsightsConfig.Band?,
        collecting: Collecting?,
        items: List<Item>,
        positives: List<RecoverySignal>,
        negatives: List<RecoverySignal>,
        load: RecoveryLoad?,
        confidence: String? = null
    ) = RecoveryResult(
        day = day, status = status, score = score, label = band?.id, labelText = band?.label,
        recommendation = band?.recommendation, confidence = confidence, collecting = collecting,
        components = items.map {
            RecoveryComponent(
                id = it.id, weight = it.weight, available = it.available, value = roundTo(it.value, 2),
                baseline = roundTo(it.baseline, 2), delta = roundTo(it.delta, 2), pct = roundTo(it.pct, 1), z = roundTo(it.z, 2),
                subscore = roundTo(it.subscore, 1), impact = roundTo(it.impact, 1), baselineN = it.baselineN,
                baselineConfidence = it.baselineConfidence, fallback = it.fallback,
                consistencyDeviationMin = roundTo(it.consistencyDeviationMin, 1), consistencySubscore = roundTo(it.consistencySubscore, 1)
            )
        },
        positives = positives, negatives = negatives, load = load
    )
}
